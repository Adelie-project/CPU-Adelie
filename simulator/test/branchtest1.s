addi r5, r0, $0   ; ants問題のような状況、アリは2匹、壁はx=0と100にある、r5は壁の左端
addi r6, r0, $100 ; r6は壁の右端
addi r7, r0, $25  ; r7はアリAの位置、最初x=25にある
addi r8, r0, $75  ; r8はアリBの位置、最初x=75にある
addi r9, r0, $7   ; r9はアリAの速度、速度だから向きもある
addi r10, r0, $-7 ; r10はアリBの速度、いま左向き
addi r11, r0, $30 ; r11はカウンタ、とりあえず30秒経過させる
addi r20, r0, $200 ; 定数
slt r12, r7, r8   ; アリAがBの左側ならr12=1
add r7, r7, r9   ;
add r8, r8, r10  ; まずアリを一秒進める、壁や向かい側のアリを貫通させても良い
slt r13, r7, r8   ; アリAがBの左側ならr13=1
bne r12, r13, $64 ; <アリどうしの位置関係が変わった場合の処理>に飛ぶ
blt r5, r7, $12   ; アリAが左側の壁にのめり込んでいる場合は以下の処理
sub r7, r5, r7
sub r9, r0, r9
blt r5, r8, $12   ; アリBが左側の壁にのめり込んでいる場合は以下の処理
sub r8, r5, r8
sub r10, r0, r10
blt r7, r6, $12   ; アリAが右側の壁にのめり込んでいる場合は以下の処理
sub r7, r20, r7
sub r9, r0, r9
blt r8, r6, $12   ; アリBが右側の壁にのめり込んでいる場合は以下の処理
sub r8, r20, r8
sub r10, r0, r10
addi r11, r11, $-1 ; カウントダウン
bne r11, r0, $-72 ; カウンタが0になるまで繰り返し、最後はr7が35,r8が65になっているはず
beq r0, r0, $0    ; ここで無限ループ
xor r7, r7, r8    ; <アリどうしの位置関係が変わった場合の処理>
xor r8, r7, r8
xor r7, r7, r8    ; まず座標をswap
sub r9, r0, r9
sub r10, r0, r10  ; 次に速度の向きを変える
beq r0, r0, $-80  ;必ず戻る
