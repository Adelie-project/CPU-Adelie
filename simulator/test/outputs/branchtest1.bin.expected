Now additional value of PC is '4'.
If this is as intended, enter 'r', others to terminate.


PC = 00000074

r00:00000000 r01:00000000 r02:00000000 r03:00000000 r04:00000000 r05:00000000 r06:00000064 r07:00000023 
r08:00000041 r09:00000007 r10:FFFFFFF9 r11:00000000 r12:00000001 r13:00000001 r14:00000000 r15:00000000 
r16:00000000 r17:00000000 r18:00000000 r19:00000000 r20:000000C8 r21:00000000 r22:00000000 r23:00000000 
r24:00000000 r25:00000000 r26:00000000 r27:00000000 r28:00000000 r29:00000000 r30:00000000 r31:00000000 

program infinitely loops at pc 116, simulation stops
